README.md
# Email Integration Pack

This pack allows integration with Email Services.

## Pre-Requisites

This pack installs Python dependencies. Your system must have a compiler installed first: - e.g. `apt install gcc`
or `yum install gcc`.

## Configuration

Copy the example configuration in [email.yaml.example](./email.yaml.example)
to `/opt/stackstorm/configs/email.yaml` and edit as required.

Configure values as described in the sections below.

**Note** : When modifying the configuration in `/opt/stackstorm/configs/` please
           remember to tell StackStorm to load these new values by running
           `st2ctl reload --register-configs`

# Actions

### send_email

When sending email the configuration option `secure` can be used to enable TLS
for the connection. `smtp_auth` is used to control if user authentication is required before
sending the message to the SMTP server.

```yaml
smtp_accounts:
  - name: me
    password: "super_S3c4e3t!"
    port: 587
    secure: true
    server: "smtp.example.com"
    smtp_auth: true
    username: "me@example.com"
  - name: you
    password: "topsecret!"
    port: 587
    secure: true
    server: "smtp.example.com"
    smtp_auth: true
    username: "you@example.com"
```

Ensure your configuration is registered:

```
st2ctl reload --register-configs
```

You can send an email from `me@example.com` to `you@example.com` using the following:

```
st2 run email.send_email account=example email_to=you@example.com message="Hi there, wow!" email_from="me@example.com" subject="Intros"
```

## Sensors
### SMTP Sensor

The SMTP Sensor runs a local server as defined in `email.yaml`, receiving email messages and
emitting triggers into the system. This server is indiscriminate with what messages are received,
and for whom. Basically, a catch-all for emails into the system.

As such, things like email filtering should happen upstream, or this should be run in a controlled
environment.

It takes these configuration options (defaults shown):

```yaml
sensor_smtp_listen_ip: '127.0.0.1'
sensor_smtp_listen_port: 1025
```

### IMAP Sensor

The IMAP Sensor logs into any number of IMAP servers as defined in `email.yaml`, polling for unread
email messages and emitting triggers into the system. Sensor looks at one account and one folder at
a time, and can be independently configured.

Typical IMAP related configuration options look like this:

```yaml
attachment_datastore_ttl: 1800
imap_accounts:
  - name: example
    download_attachments: true
    folder: "INBOX"
    password: "super_S3c4e3t!"
    port: 993
    secure: true
    server: "imap.example.com"
    username: "me@example.com"
  - name: example2
    download_attachments: true
    folder: "INBOX"
    password: "topsecret!"
    port: 993
    secure: true
    server: "imap.example2.com"
    username: "me@example2.com"
max_attachment_size: 1024
```

The following attachment settings can be configured:

* ``attachment_datastore_ttl`` - TTL in seconds for the attachment value which is
  stored in the datastore.
* ``download_attachments`` - True to download the attachment and store them in the datastore
* ``max_attachment_size`` - Maximum size of download attachment bytes. If an
  attachment exceeds this size the attachment won't be stored in the datastore.

If ``download_attachments`` attribute for a particular IMAP server is set to ``True``,
attachments will be automatically downloaded and stored in the built-in datastore under
a unique key. This key will be available in the trigger payload (see the trigger example
below) so you can retrieve those attachments later (e.g. inside an action).

By default, those values have a TTL of 30 minutes which means they will be automatically removed
from the datastore after 30 minutes.

Keep in mind that all the attachments which contain binary (non plain-text, mime-type
!= text/plain) data are base64 encoded before they are stored in the datastore.

### email.imap.message trigger

Example trigger payload:

```json
{
    "uid": 5780,
    "message_id": "<CAJMHEmJs_5hO_PS9huzTAOv60xkTvYz7ETd1=WXFdpa-bzGpUA@mail.gmail.com>",
    "from": "Tomaž Muraus<tomaz@tomaz.me>",
    "to": "Tomaz Muraus <tomaz.muraus@gmail.com>",
    "subject": "test email with attachment",
    "body": "hello from stackstorm!\n",
    "headers": [
        [
            "Delivered-To",
            "tomaz.muraus@gmail.com"
        ],
        [
            "Received",
            "by 10.182.33.5 with SMTP id n5csp2470866obi;        Wed, 10 Jun 2015 00:01:41 -0700 (PDT)"
        ],
        [
            "X-Received",
            "by 10.50.61.130 with SMTP id p2mr3867507igr.9.1433919701526;        Wed, 10 Jun 2015 00:01:41 -0700 (PDT)"
        ],
        [
            "Return-Path",
            "<kami@k5-storitve.net>"
        ],
        [
            "Received",
            "from mail-ig0-x230.google.com (mail-ig0-x230.google.com. [2607:f8b0:4001:c05::230])        by mx.google.com with ESMTPS id we7si8228902icb.8.2015.06.10.00.01.41        for <tomaz.muraus@gmail.com>        (version=TLSv1.2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);        Wed, 10 Jun 2015 00:01:41 -0700 (PDT)"
        ],
        [
            "Received-Spf",
            "pass (google.com: domain of kami@k5-storitve.net designates 2607:f8b0:4001:c05::230 as permitted sender) client-ip=2607:f8b0:4001:c05::230;"
        ],
        [
            "Authentication-Results",
            "mx.google.com;       spf=pass (google.com: domain of kami@k5-storitve.net designates 2607:f8b0:4001:c05::230 as permitted sender) smtp.mail=kami@k5-storitve.net;       dkim=pass header.i=@tomaz.me"
        ],
        [
            "Received",
            "by mail-ig0-x230.google.com with SMTP id pi8so28796173igb.0        for <tomaz.muraus@gmail.com>; Wed, 10 Jun 2015 00:01:41 -0700 (PDT)"
        ],
        [
            "Dkim-Signature",
            "v=1; a=rsa-sha256; c=relaxed/relaxed;        d=k5-storitve.net; s=google;        h=mime-version:sender:from:date:message-id:subject:to:content-type;        bh=dnyggWfMbPP+DPkMG3PmSW5Y7wvt84XbnBhbgnAUusg=;        b=VmO+M+kBxVU7BwCzreI3vza5kvkxUwkCsiZrlunnMfMnP60RJBJHhE3HtQmIITkjoD         v5fAou2vcSIm5eY/CYAbSJyzzhP6sNbVoHJl1Q90Gqb1KA8g3+hF+mBOBhIqEf0fKiRt         07f0maRrvwJdI54HHRuroE7jSs8DHNllWBJfY="
        ],
        [
            "Dkim-Signature",
            "v=1; a=rsa-sha256; c=relaxed/relaxed;        d=tomaz.me; s=google;        h=mime-version:sender:from:date:message-id:subject:to:content-type;        bh=dnyggWfMbPP+DPkMG3PmSW5Y7wvt84XbnBhbgnAUusg=;        b=YQGGwcMru9HaWMTbcEtkDdALkSLwEANo/ruZ76REaeW8Hnj0U6aM+MLLRKLsiFwSM+         THzY92cpVDAlYkbDLyqN+PctHyOx3ofRobRjjv2741SzV8ZTYLPSyaqsLtOJlRbfo16m         U+9vVgux9/xGrGQnF4DckO86DlcDPPL4oPgBI="
        ],
        [
            "X-Google-Dkim-Signature",
            "v=1; a=rsa-sha256; c=relaxed/relaxed;        d=1e100.net; s=20130820;        h=x-gm-message-state:mime-version:sender:from:date:message-id:subject         :to:content-type;        bh=dnyggWfMbPP+DPkMG3PmSW5Y7wvt84XbnBhbgnAUusg=;        b=TEvRNj86wdtz8SQWp1TfqIYOyFbVh6aEhVWcO1hXFf26fh6M36pRTty48qDzCN7dZb         hHZuKLTneCcgnhnt6bbVgR23AkeMtujFc2QGawF3e/So6Z8VGc1VMBoCdd3li0Epqj+w         OisxHlzV5HNhkcj+UB77345yGZBapcgoZxtn5/m6OaL+wDlWjLfgu0j0FHiMDlftJgF3         yMkgONFIZyVqz1xOey4rvjNBNpGowfF9ei0r869PzUjVLtYuw2UuvhXn0AbduxQHMxmA         1QY67srbOsz5DR+u0bX+2euzI7s5KDFCh1hredYctdr87lyEhJew6HYfCNYUXxxLF5R+         h8Wg=="
        ],
        [
            "X-Gm-Message-State",
            "ALoCoQmCaIs0LyyKWSiGNbUHv6N6osKVmVAuQ7zqUeNA+/7uh8fopFqq/hoF6Fry25ZELwjwbEPr"
        ],
        [
            "X-Received",
            "by 10.42.203.4 with SMTP id fg4mr3192769icb.52.1433919701095; Wed, 10 Jun 2015 00:01:41 -0700 (PDT)"
        ],
        [
            "Mime-Version",
            "1.0"
        ],
        [
            "Sender",
            "kami@k5-storitve.net"
        ],
        [
            "Received",
            "by 10.50.43.66 with HTTP; Wed, 10 Jun 2015 00:01:20 -0700 (PDT)"
        ],
        [
            "From",
            "Tomaž Muraus <tomaz@tomaz.me>"
        ],
        [
            "Date",
            "Wed, 10 Jun 2015 15:01:20 +0800"
        ],
        [
            "X-Google-Sender-Auth",
            "WfbOSpNXpbHPX2uAz0lP7j-8z7g"
        ],
        [
            "Message-Id",
            "<CAJMHEmJs_5hO_PS9huzTAOv60xkTvYz7ETd1=WXFdpa-bzGpUA@mail.gmail.com>"
        ],
        [
            "Subject",
            "test email with attachment"
        ],
        [
            "To",
            "Tomaz Muraus <tomaz.muraus@gmail.com>"
        ],
        [
            "Content-Type",
            [
                "multipart/mixed",
                {
                    "boundary": "20cf303f64066bc8ff0518247248"
                }
            ]
        ]
    ],
    "date": "Wed, 10 Jun 2015 15:01:20 +0800",
    "has_attachments": true,
    "attachments": [
        {
            "file_name": "hello stackstorm.txt",
            "datastore_key": "attachments-0d61eabdb749789c96853dcbbd933884",
            "content_type": "text/plain"
        }
    ],
    "mailbox_metadata": {
        "server": "email.stackstorm.com",
        "port": 993,
        "user": "kami@stackstorm.com",
        "folder": "Process",
        "ssl": true
    }
}
```

### Troubleshooting

If you are trying to use Gmail (e.g. imap.gmail.com or smtp.gmail.com with a Gmail or Google apps account):

* For IMAP and SMTP use an application specific password
* For IMAP start with port 587. Currently this is the only port that will give you errors and not just close the connection. 


CHANGES.md
# Change Log

# 2.0.3

- Set smtp username and password as not required. 

# 2.0.2

- Add date to header in the SendEmail action so the spam score will be lower

# 2.0.1

- Fix Python 3 related issue in IMAPSensor

# 2.0.0

* Drop Python 2.7 support

# 1.1.6

- Use the correct variable for _attachment_datastore_ttl in imap_sensor.
# 1.1.5

- Add explicit support for Python 2 and 3

# 1.1.4

- Pin tld dependency to <0.11 for Python < 3.0

# 1.1.3

- Minor linting fixes

# 1.1.1

- Version bump to fix tagging issues

# 1.1.0

- Updated to flanker >= 0.9.0
- Updated send_email.py to support Unicode in subject and message fields

# 1.0.3

- Add `attachments` parameter to the send_email action. Expected a list of file paths if given, to be attached to the email.

# 1.0.2

- Add `mime` parameter to the send_email action.  If the value is `html`, the mime type of the email will be `text/html`.  Otherwise, the mime will be `text/plain`.

# 1.0.0

- Replaced 'imap_mailboxes' with 'imap_accounts' and 'smtp_accounts'. Replaced 'smtp_list_ip'
  and 'smtp_listen_port' with 'sensor_smtp_listen_ip' and 'sensor_smtp_listen_port' respectively.
  You must update any existing email.yaml file.

# 0.4.0

- Changed `email_to` data type from string to list in order to support multiple email address
  destinations.

# 0.3.0

- Updated action `runner_type` from `run-python` to `python-script`

# 0.2.0

- Rename `config.yaml` to `config.schema.yaml` and update to use schema.

# 0.1.0

- First release 


actions/send_email.py
import os
from st2common.runners.base_action import Action
from smtplib import SMTP
from email.header import Header
from email.mime.application import MIMEApplication
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from email.utils import formatdate


class SendEmail(Action):
    def run(self, email_from, email_to, subject, message,
            account, email_cc=None, mime="plain", attachments=None):

        if mime not in ['plain', 'html']:
            raise ValueError('Invalid mime provided: ' + mime)

        accounts = self.config.get('smtp_accounts', None)
        if accounts is None:
            raise ValueError('"smtp_accounts" config value is required to send email.')
        if not accounts:
            raise ValueError('at least one account is required to send email.')

        try:
            kv = {}
            for a in accounts:
                kv[a['name']] = a
            account_data = kv[account]
        except KeyError:
            raise KeyError('The account "{}" does not seem to appear in the configuration. '
                           'Available accounts are: {}'.format(account, ",".join(kv.keys())))

        msg = MIMEMultipart()
        msg['Subject'] = Header(subject, 'utf-8')
        msg['From'] = email_from
        msg['To'] = ", ".join(email_to)
        msg['Date'] = formatdate(localtime=True)
        msg.attach(MIMEText(message, mime, 'utf-8'))
        if email_cc is None:
            email_cc = []

        if email_cc:
            msg['Cc'] = ", ".join(email_cc)
            email_to += email_cc

        attachments = attachments or tuple()
        for filepath in attachments:
            filename = os.path.basename(filepath)
            with open(filepath, 'rb') as f:
                part = MIMEApplication(f.read(), Name=filename)
            part['Content-Disposition'] = 'attachment; filename="{}"'.format(filename)
            msg.attach(part)

        s = SMTP(account_data['server'], int(account_data['port']), timeout=20)
        s.ehlo()
        if account_data.get('secure', True) is True:
            s.starttls()
        if account_data.get('smtp_auth', True) is True:
            s.login(account_data.get('username', str()), account_data.get('password', str()))
        s.sendmail(email_from, email_to, msg.as_string())
        s.quit()
        return


sensors/smtp_sensor.py
import eventlet

eventlet.monkey_patch(
    os=True,
    select=True,
    socket=True,
    thread=True,
    time=True)

from eventlet.green import asyncore
import smtpd_green as smtpd
from flanker import mime
import hashlib
import base64
import six

from st2reactor.sensor.base import Sensor


class SMTPSensor(Sensor):
    def __init__(self, sensor_service, config=None):
        super(SMTPSensor, self).__init__(sensor_service=sensor_service,
                                         config=config)

        self._trigger = 'email.smtp.message'
        self._logger = self._sensor_service.get_logger(__name__)
        self._server = None
        self._listen_ip = self._config.get('sensor_smtp_listen_ip', '127.0.0.1')
        self._listen_port = self._config.get('sensor_smtp_listen_port', 1025)

    def setup(self):
        self._logger.debug('[SMTPSensor]: entering setup')
        self._server = St2SMTPServer(localaddr=(self._listen_ip, self._listen_port),
                                     remoteaddr=None,
                                     sensor_service=self._sensor_service,
                                     logger=self._logger,
                                     trigger=self._trigger)

    def run(self):
        self._logger.debug('[SMTPSensor]: entering run')
        asyncore.loop()  # pylint: disable=no-member

    def cleanup(self):
        self._logger.debug('[SMTPSensor]: entering cleanup')

    def add_trigger(self, trigger):
        pass

    def update_trigger(self, trigger):
        pass

    def remove_trigger(self, trigger):
        pass


class St2SMTPServer(smtpd.SMTPServer):  # pylint: disable=no-member
    def __init__(self, localaddr, remoteaddr, sensor_service, logger, trigger):
        smtpd.SMTPServer.__init__(self, localaddr, remoteaddr)  # pylint: disable=no-member
        self._logger = logger
        self._trigger = trigger
        self._sensor_service = sensor_service

    def _flattern_headers(self, headers):
        # Flattern headers and make sure they only contain simple types so they
        # can be serialized in a trigger
        result = []

        for pair in headers:
            name = pair[0]
            value = pair[1]

            if not isinstance(value, six.string_types):
                value = str(value)

            result.append([name, value])

        return result

    def process_message(self, peer, mailfrom, rcpttos, data):
        self._logger.debug('posting message from {} to {}'.format(mailfrom, rcpttos))

        message = self.parse_message(mailfrom, rcpttos, data)
        self._sensor_service.dispatch(trigger=self._trigger, payload=message)
        return

    def parse_message(self, mailfrom, rcpttos, data):
        message = mime.from_string(data)
        headers = message.headers.items()
        headers = self._flattern_headers(headers=headers)
        payload = {
            'from': None,
            'to': None,
            'subject': None,
            'date': None,
            'body_plain': None,
            'body_html': None,
            'attachments': [],
            'headers': headers,
        }

        # Try to get the addressee via headers, or
        # fall-back to raw protocol request
        if 'To' in message.headers.keys():
            payload['to'] = message.headers['To']
        else:
            payload['to'] = rcpttos

        # Try to get the recipient via headers, or
        # fall-back to raw protocol request
        if 'From' in message.headers.keys():
            payload['from'] = message.headers['From']
        else:
            payload['from'] = mailfrom

        if 'Subject' in message.headers.keys():
            payload['subject'] = message.headers['Subject']

        if 'Date' in message.headers.keys():
            payload['date'] = message.headers['Date']

        # Body
        if message.content_type.is_singlepart():
            payload['body_plain'] = message.body
        elif message.content_type.is_multipart():
            for part in message.parts:
                content_type = part.content_type[0]

                if content_type == 'text/plain':
                    payload['body_plain'] = part.body
                elif content_type == 'text/html':
                    payload['body_html'] = part.body
                elif part.is_attachment():
                    attachment = {
                        'filename': part.detected_file_name,
                        'md5': hashlib.md5(part.body).hexdigest(),
                        'sha1': hashlib.sha1(part.body).hexdigest(),
                        'data': base64.b64encode(part.body),
                        'encoding': part.content_encoding[0],
                        'type': content_type,
                    }
                    payload['attachments'].append(attachment)

        return payload


sensors/imap_sensor.py
import hashlib
import base64

import six
import eventlet
import easyimap
from flanker import mime

from st2reactor.sensor.base import PollingSensor

__all__ = [
    'IMAPSensor'
]

eventlet.monkey_patch(
    os=True,
    select=True,
    socket=True,
    thread=True,
    time=True)

DEFAULT_DOWNLOAD_ATTACHMENTS = False
DEFAULT_MAX_ATTACHMENT_SIZE = 1024
DEFAULT_ATTACHMENT_DATASTORE_TTL = 1800
DEFAULT_POLLING_INTERVAL = 30


class IMAPSensor(PollingSensor):
    def __init__(self, sensor_service, config=None, poll_interval=DEFAULT_POLLING_INTERVAL):
        super(IMAPSensor, self).__init__(sensor_service=sensor_service,
                                         config=config,
                                         poll_interval=self._config.get('email_polling_interval',
                                                                        poll_interval))

        self._trigger = 'email.imap.message'
        self._logger = self._sensor_service.get_logger(__name__)

        self._max_attachment_size = self._config.get('max_attachment_size',
                                                     DEFAULT_MAX_ATTACHMENT_SIZE)
        self._attachment_datastore_ttl = self._config.get('attachment_datastore_ttl',
                                                          DEFAULT_ATTACHMENT_DATASTORE_TTL)
        self._accounts = {}

    def setup(self):
        self._logger.debug('[IMAPSensor]: entering setup')

    def poll(self):
        self._logger.debug('[IMAPSensor]: entering poll')

        if 'imap_accounts' in self._config:
            self._parse_accounts(self._config['imap_accounts'])

        for name, values in self._accounts.items():
            mailbox = values['connection']
            download_attachments = values['download_attachments']
            mailbox_metadata = values['mailbox_metadata']

            self._poll_for_unread_messages(name=name, mailbox=mailbox,
                                           download_attachments=download_attachments,
                                           mailbox_metadata=mailbox_metadata)
            mailbox.quit()

    def cleanup(self):
        self._logger.debug('[IMAPSensor]: entering cleanup')

        for name, values in self._accounts.items():
            mailbox = values['connection']
            self._logger.debug('[IMAPSensor]: Disconnecting from {0}'.format(name))
            mailbox.quit()

    def add_trigger(self, trigger):
        pass

    def update_trigger(self, trigger):
        pass

    def remove_trigger(self, trigger):
        pass

    def _parse_accounts(self, accounts):
        for config in accounts:
            mailbox = config.get('name', None)
            server = config.get('server', 'localhost')
            port = config.get('port', 143)
            user = config.get('username', None)
            password = config.get('password', None)
            folder = config.get('folder', 'INBOX')
            ssl = config.get('secure', False)
            download_attachments = config.get('download_attachments', DEFAULT_DOWNLOAD_ATTACHMENTS)

            if not user or not password:
                self._logger.debug("""[IMAPSensor]: Missing
                    username/password for {0}""".format(mailbox))
                continue

            if not server:
                self._logger.debug("""[IMAPSensor]: Missing server
                    for {0}""".format(mailbox))
                continue

            try:
                connection = easyimap.connect(server, user, password,
                                              folder, ssl=ssl, port=port)
            except Exception as e:
                message = 'Failed to connect to mailbox "%s": %s' % (mailbox, str(e))
                raise Exception(message)

            item = {
                'connection': connection,
                'download_attachments': download_attachments,
                'mailbox_metadata': {
                    'server': server,
                    'port': port,
                    'user': user,
                    'folder': folder,
                    'ssl': ssl
                }
            }
            self._accounts[mailbox] = item

    def _poll_for_unread_messages(self, name, mailbox, mailbox_metadata,
                                  download_attachments=False):
        self._logger.debug('[IMAPSensor]: polling mailbox {0}'.format(name))

        messages = mailbox.unseen()

        self._logger.debug('[IMAPSensor]: Processing {0} new messages'.format(len(messages)))
        for message in messages:
            self._process_message(uid=message.uid, mailbox=mailbox,
                                  download_attachments=download_attachments,
                                  mailbox_metadata=mailbox_metadata)

    def _process_message(self, uid, mailbox, mailbox_metadata,
                         download_attachments=DEFAULT_DOWNLOAD_ATTACHMENTS):
        message = mailbox.mail(str(uid), include_raw=True)
        mime_msg = mime.from_string(message.raw)

        body = message.body
        sent_from = message.from_addr
        sent_to = message.to
        subject = message.title
        date = message.date
        message_id = message.message_id
        headers = mime_msg.headers.items()
        has_attachments = bool(message.attachments)

        # Flatten the headers so they can be unpickled
        headers = self._flattern_headers(headers=headers)

        payload = {
            'uid': uid,
            'from': sent_from,
            'to': sent_to,
            'headers': headers,
            'date': date,
            'subject': subject,
            'message_id': message_id,
            'body': body,
            'has_attachments': has_attachments,
            'attachments': [],
            'mailbox_metadata': mailbox_metadata
        }

        if has_attachments and download_attachments:
            self._logger.debug('[IMAPSensor]: Downloading attachments for message {}'.format(uid))
            result = self._download_and_store_message_attachments(message=message)
            payload['attachments'] = result

        self._sensor_service.dispatch(trigger=self._trigger, payload=payload)

    def _download_and_store_message_attachments(self, message):
        """
        Method which downloads the provided message attachments and stores them in a datasatore.

        :rtype: ``list`` of ``dict``
        """
        attachments = message.attachments

        result = []
        for (file_name, content, content_type) in attachments:
            attachment_size = len(content)

            if len(content) > self._max_attachment_size:
                self._logger.debug(('[IMAPSensor]: Skipping attachment "{}" since its bigger '
                                    'than maximum allowed size ({})'.format(file_name,
                                                                            attachment_size)))
                continue

            datastore_key = self._get_attachment_datastore_key(message=message,
                                                               file_name=file_name)

            # Store attachment in the datastore
            if content_type == 'text/plain':
                value = content
            else:
                value = base64.b64encode(content)

            self._sensor_service.set_value(name=datastore_key, value=value,
                                           ttl=self._attachment_datastore_ttl,
                                           local=False)
            item = {
                'file_name': file_name,
                'content_type': content_type,
                'datastore_key': datastore_key
            }
            result.append(item)

        return result

    def _get_attachment_datastore_key(self, message, file_name):
        key = '%s-%s' % (message.uid, file_name)
        key = 'attachments-%s' % (hashlib.md5(key).hexdigest())
        return key

    def _flattern_headers(self, headers):
        # Flattern headers and make sure they only contain simple types so they
        # can be serialized in a trigger
        result = []

        for pair in headers:
            name = pair[0]
            value = pair[1]

            if not isinstance(value, six.string_types):
                value = str(value)

            result.append([name, value])

        return result
